### DISCUSSION TITLE
Cannot build latest HEAD of hubs-cloud due to "worker-loader" error

### DISCUSSION THREAD

**Discussion by MegaMotion** [mozilla/hubs#3680] [Hubs Cloud Questions] [answered]
Hi, not sure what went wrong for me here, but after working on a custom client branch for a little while, I decided to switch back to the main hubs-cloud branch and pull down recent updates. However, it failed to build, and when I started a fresh clone it also failed on the same error. My npm is v12.16.3, running windows 10.  

The problem appears to be related to something called the worker factory, or worker-loader?  

`  
D:\HubsEvents\hubs2\node_modules\worker-loader\dist\index.js:121
      worker.factory = (0, _workers2.default)(worker.file, compilation.assets[worker.file].source(), options);
                                                                                           ^
  TypeError: Cannot read property 'source' of undefined at D:\HubsEvents\hubs2\node_modules\worker-loader\dist\index.js:121:92
`

EDIT: okay, after trying to go back to my original branch, I still have this issue, so it seems like a local problem, but ...? I tried installing stripe API tools in my custom branch, I had it starting to work there before I tried switching back to the hubs-cloud branch.  I'm from C++ game dev, still pretty unfamiliar with node JS environments, sorry.

**Comment by MegaMotion** [comment]
Well, still in the dark about what's broken, but on a second machine it seems to be working, so, lucky me. :-)

**Reply by robinkwilson** [reply]
Check the `node -v` and `npm -v`. A node version of 12.16.3 seems very low. I'd try upgrading to 14.

**Reply by MegaMotion** [reply]
Thanks, Robin! And, whew, that was a struggle. Turns out I jumped the gun earlier, my second machine got past the error I was having on the first one but crashed later due to something else.

After fighting with it on two machines all morning though, and installing brand new node 14.15.4 on both of them, I finally got the first one to work by running the console as admin - I don't remember that being a requirement, did I miss something? Or was it just necessary this one time in order to let it create a directory or some such?

**Reply by robinkwilson** [reply]
Oh interesting! i guess there was some dependency that needed your admin access. 

If we see this pop up more often, we'll update the hubs docs to include that as a requirement. 

Do you know the error you saw? -- or the one after you updated node to 14.15.4

**Reply by MegaMotion** [reply]
Ah, sorry, I appear to have not taken good notes at that point. I know I saw this featuring prominently, although it's not very useful by itself:

 Error: EPERM: operation not permitted

But sadly I don't remember what file it was choking on.

**Reply by MegaMotion** [reply]
Ah, well, turns out this wasn't as fixed as I thought... I've been coding along happily for days, but just today the bug came back with a vengeance, affecting both the code I was working on (which hadn't really changed significantly since the last time it was running fine) and the base hubs-cloud branch which is straight from the horse's mouth.

I do have some error logging data this time, however:

`> hubs@0.0.1 build D:\HubsEvents\hubs3
> rimraf ./dist && webpack --mode=production

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db
D:\HubsEvents\hubs3\node_modules\worker-loader\dist\index.js:121
      worker.factory = (0, _workers2.default)(worker.file, compilation.assets[worker.file].source(), options);
                                                                                           ^

TypeError: Cannot read property 'source' of undefined
    at D:\HubsEvents\hubs3\node_modules\worker-loader\dist\index.js:121:92
    at D:\HubsEvents\hubs3\node_modules\webpack\lib\Compiler.js:343:11
    at D:\HubsEvents\hubs3\node_modules\webpack\lib\Compiler.js:681:15
`

Any help here would be *much* appreciated, it is a very nasty and completely show stopping bug. :-(

**Comment by robinkwilson** [comment, marked as answer]
Sounds like a nasty bug. Unfortunately I'm not sure where it went wrong. I get the caniuse-lite log too (I've upgraded it, but it failed to build and deploy afterward, so I can't recommend updating that!).

Is this erroring when you run `npm run deploy` or build or `npm run dev` or `npm start`?

**Reply by MegaMotion** [reply]
It errors out when I run any of the build options - build, dev, deploy, start.

**Reply by robinkwilson** [reply]
Marking as answered for now, let us know if you have any additional specific questions/info then "unmark" to get it back in the pipeline!

**Reply by robinkwilson** [reply]
I believe when we discussed this: we ran `npm ci` and it magically got fixed. Sometimes I need to run `npm add fsevents` after (will need to add this to the packages so it's not manual) for my mac to get everything to work properly.

**Reply by MegaMotion** [reply]
Thanks, Robin! 

I just had the same thing happen again, but clearing node_modules and running npm ci again fixed it... it's really weird and I hope it goes away at some point, it seems related to the ammo physics stuff. The library it is always unable to find is "ammo-debug-drawer".

**Reply by MegaMotion** [reply]
For anyone others who find themselves having this problem... I have definitely determined that running my console window as administrator can be necessary. I just had the issue again, without admin privileges, and then restarted the console with admin privileges and it worked.
